File: backend/src/HobbyTracker.Api/Domain/Anime.cs
namespace HobbyTracker.Api.Domain;

/// <summary>
/// Anime-specific detail for a <see cref="Media"/> row. Mapped Table-Per-Type into an `anime`
/// table whose primary key (`media_id`) doubles as its foreign key to `media`, exactly as
/// <see cref="Game"/>, <see cref="Movie"/> and <see cref="TvShow"/> are.
///
/// The fourth detail table, and the one that most resembles the third without being it. An
/// anime and a show are both episodic — and this carries a second title, a source material, a
/// studio and a mean score that no show has any use for, while having **no seasons table at
/// all**. That absence is the decision: MAL numbers each cour as its own entry, so
/// <c>Sousou no Frieren</c> and <c>Sousou no Frieren 2nd Season</c> are two ids and therefore
/// two cards. <c>tv_seasons</c> exists because TMDB models a series as one title with many
/// seasons; there is nothing here to hang a child table off.
///
/// See <c>docs/anime-mal.md</c>.
/// </summary>
public class Anime : Media
{
    /// <summary>
    /// What it is called in English, from MAL's `alternative_titles.en`.
    ///
    /// <see cref="Media.Title"/> holds the romaji — <c>Sousou no Frieren</c> — because that is
    /// what MAL's own `title` is and what its search matches on, so everything the platform
    /// already does with a title needs no special case. This is the second line under a card's
    /// heading, and <b>null is the ordinary case rather than an error</b>: MAL leaves `en` absent
    /// on a great many entries, and the card then renders nothing rather than an empty line.
    ///
    /// Not `english_title` on the board row it reaches — see <c>LibraryItemDto.Subtitle</c>.
    /// The column can be specific because the column is this hobby's; the row's field is the
    /// platform's, and the next hobby to want one may not mean English by it.
    /// </summary>
    public string? EnglishTitle { get; set; }

    /// <summary>
    /// What MAL calls the shape of it: `tv`, `movie`, `ova`, `ona`, `special`, `music`.
    ///
    /// Everything MAL calls anime is on the anime board, which is what the catalogue hands you
    /// anyway — so this is a fact to print rather than a filter to apply. MAL's own word, stored
    /// verbatim and made readable on the client, exactly as <see cref="TvShow.AirStatus"/> is.
    /// </summary>
    public string? MediaType { get; set; }

    /// <summary>
    /// How many episodes the cour has.
    ///
    /// <b>MAL answers 0 for an entry that has not aired, and 0 means unknown rather than none</b>
    /// — a cour announced for 2027 is the ordinary case for it. Null is what says that
    /// truthfully here, and `ck_anime_counts_positive` is what makes forgetting the mapping fail
    /// loudly rather than let a card claim a season with no episodes in it.
    /// </summary>
    public int? EpisodeCount { get; set; }

    /// <summary>
    /// How long one episode runs, in <b>seconds</b> — MAL's own unit, from
    /// `average_episode_duration`.
    ///
    /// Kept in the unit it arrives in rather than converted on the way in. <c>tv_shows</c> stores
    /// minutes because TMDB states minutes; this states seconds because MAL states seconds, and
    /// a column that silently held a different unit from its provider is the kind of thing only
    /// ever noticed as a runtime sixty times out. The conversion happens once, in the generated
    /// column below, where Postgres does it.
    /// </summary>
    public int? EpisodeRuntimeSeconds { get; set; }

    /// <summary>
    /// The whole cour in minutes, and what a card's badge and <c>sort=length</c> both read
    /// through <c>LibraryItemDto.LengthHours</c>.
    ///
    /// <b>A Postgres generated column — the database multiplies and converts it, and nothing
    /// here ever writes it.</b> <see cref="TvShow.TotalRuntimeMinutes"/>'s argument exactly, with
    /// one more reason on top of it: this sum carries a unit conversion, so the factor of sixty
    /// lives in one place rather than in each of the three <c>LibraryService</c> readers.
    ///
    /// Null propagates for free: either factor missing makes the product null, and a null length
    /// already sorts last rather than as though nobody having timed a cour meant it took no time.
    /// </summary>
    public int? TotalRuntimeMinutes { get; private set; }

    /// <summary>
    /// Which of the four broadcast seasons it began in — `winter`, `spring`, `summer`, `fall`.
    ///
    /// MAL's own word, and worth storing beside the year rather than folded into it: anime is
    /// talked about by cour, and "Fall 2023" is how anybody would say when Frieren started.
    /// </summary>
    public string? StartSeason { get; set; }

    /// <summary>
    /// The year it began airing, from `start_season.year`.
    ///
    /// Read as the year MAL states and nothing more — a fact about the title rather than a
    /// moment, so it never goes through the journal clock, exactly as
    /// <see cref="TvShow.FirstAirYear"/> and <see cref="Movie.ReleaseYear"/> are.
    /// </summary>
    public int? StartYear { get; set; }

    /// <summary>
    /// Where the title itself stands, from MAL's `status`: `finished_airing`,
    /// `currently_airing`, `not_yet_aired`.
    ///
    /// <b>Named `air_status` rather than `status`</b>, for <see cref="TvShow.AirStatus"/>'s
    /// reason: <c>log_entries.status</c> is where *you* are and is a closed four-value
    /// vocabulary shared by every hobby, and two columns called `status` on tables joined in
    /// every board query is a mistake waiting for the first person who types the shorter name.
    /// </summary>
    public string? AirStatus { get; set; }

    /// <summary>
    /// What it was adapted from, from MAL's `source`: `manga`, `light_novel`, `original`,
    /// `visual_novel`, `game`, and a dozen more.
    ///
    /// <b>Named `source_material` rather than `source`, and that is the same collision again.</b>
    /// <see cref="Media.SourceId"/> is which provider a row came from and is a foreign key into
    /// a lookup table; this is what the story was adapted from and is MAL's free text. One of
    /// them is a fact about the record and the other is a fact about the work.
    ///
    /// A genuinely good fact for the drawer: whether a thing is an original or an adaptation is
    /// most of what anybody wants to know about it before starting.
    /// </summary>
    public string? SourceMaterial { get; set; }

    /// <summary>
    /// MAL's genres for this title, as a Postgres text[].
    ///
    /// <b>MAL flattens three taxonomies into one array</b> — genres, themes and demographics all
    /// arrive in `genres`. Frieren comes back as Adventure, Award Winning, Drama, Fantasy,
    /// Shounen: two genres, a MAL tag and a demographic in one list. So the ordered
    /// specific-before-generic list this codebase already uses works on it directly, with no
    /// merging step — <c>frontend/src/hobbies/anime.ts</c> puts Isekai and Iyashikei above
    /// Action and Drama exactly as Visual Novel sits above Role-playing (RPG).
    /// </summary>
    public List<string> Genres { get; set; } = [];

    /// <summary>
    /// The genre chosen to stand for this title, overriding the automatic pick. Null means "use
    /// the automatic one", not "no genre" — and a MAL refresh leaves it alone, exactly as it
    /// leaves the other three hobbies' alone.
    /// </summary>
    public string? PrimaryGenre { get; set; }

    /// <summary>
    /// Who animated it, from MAL's `studios`.
    ///
    /// The byline under the drawer's title — a game's developers, a film's directors and a
    /// show's creators all answer the same question, and for anime the answer anybody gives is
    /// the studio. An array because co-productions are common.
    /// </summary>
    public List<string> Studios { get; set; } = [];

    /// <summary>
    /// What MAL's own users score it, out of ten.
    ///
    /// The same scale this app's ratings use, which is the whole reason it is worth carrying:
    /// the drawer can print it beside your own without either number needing a footnote.
    /// `ck_anime_mean_score_range` holds the scale, so a figure outside it fails as the mapping
    /// fault it would be rather than sitting next to a rating it cannot be compared with.
    /// </summary>
    public decimal? MeanScore { get; set; }
}

File: frontend/src/hobbies/anime.ts
import { getAnime, searchAnime, setAnimeGenre } from '../api/anime';
import { formatHours, formatMinutes } from '../lib/hours';
import type { HobbyDefinition, TitleFact } from './types';

/**
 * Anime: MyAnimeList, and the first hobby that is not television.
 *
 * It looks like the third file and is not. A show and an anime are both episodic, but MAL
 * models each **cour** as its own entry — `Sousou no Frieren` and `Sousou no Frieren 2nd
 * Season` are two ids and therefore two cards — where TMDB models a series as one title with
 * many seasons. That single fact is where every difference below comes from: the pass has an
 * episode and no season, the card carries a second title, and there is no season list anywhere.
 *
 * It also sits beside television deliberately rather than inside it. The decision was the
 * user's: *"Anime is a very specific genre that should be separated"* — the distinction is in
 * the titles, not the mechanics, and the mechanics being identical is not an argument for
 * merging them. `TmdbClient` keeps anime off the TV board to match.
 */
export const ANIME: HobbyDefinition = {
  slug: 'anime',

  // Watching and Watched, as films and shows are. A fourth word for the same act would read as
  // a distinction where there is none — and the nav puts all three side by side.
  columnLabel: {
    Backlog: 'Backlog',
    InProgress: 'Watching',
    Completed: 'Watched',
    Dropped: 'Dropped',
  },

  // Television's word, for television's reason: this board has two durations — the whole cour,
  // which is what a card's badge and `sort=length` mean, and one episode, which is what the
  // drawer's facts band means. Calling both Runtime would put two controls a few inches apart
  // naming numbers a factor of twenty-eight apart.
  lengthLabel: 'Time to watch',

  // The tilde, as television has it. MAL states an *average* episode duration and Postgres
  // multiplies it, so this is an estimate in exactly the way a film's exact runtime is not.
  formatLength: (hours) => `~${formatHours(hours)}`,
  describeLength: (hours) => `About ${hours} hours to watch`,

  countPasses: (count) => `${count} watches`,

  describeRemoval: (title, entryCount) =>
    entryCount > 1
      ? `Takes ${title} off your board — all ${entryCount} watches, and their notes.`
      : `Takes ${title} off your board.`,

  /**
   * MAL's vocabulary, which is three taxonomies in one array.
   *
   * **This is the good news of the provider.** MAL flattens genres, themes and demographics
   * into a single `genres` list — Frieren comes back as Adventure, Award Winning, Drama,
   * Fantasy, Shounen — so the ordered specific-before-generic rule this codebase already uses
   * works on it directly, with no merging step. The four themes at the top are the ones that
   * describe an evening rather than a subject, which is Visual Novel's placement argument
   * applied to a catalogue that has far more of them.
   *
   * Thirteen, where games has eleven and the two TMDB lists have ten. Anime genuinely carries
   * more axes that mean different evenings: Isekai and Mecha and Iyashikei have no counterpart
   * on any other board, and cutting to ten would have to lose one of Sci-Fi or Fantasy, which
   * between them cover most of the catalogue.
   *
   * **Demographics are deliberately unpainted.** Shounen, Seinen, Shoujo and Josei say who a
   * thing was published for rather than what watching it is like, and a title carrying nothing
   * else then paints as nothing — which is honest.
   *
   * **Six hues are the films palette's**, on television's shared-vocabulary argument exactly: a
   * horror is the same evening on either board, and a year-in-review page is the one screen
   * that will put two hobbies together. Only the seven words anime owns outright are minted.
   * The floor is Mecha against Isekai at 0.103; every other pair clears 0.107.
   */
  genres: [
    { name: 'Isekai', stripe: 'bg-genre-isekai' },
    { name: 'Iyashikei', stripe: 'bg-genre-iyashikei' },
    { name: 'Mecha', stripe: 'bg-genre-mecha' },
    { name: 'Psychological', stripe: 'bg-genre-psychological' },
    { name: 'Sports', stripe: 'bg-genre-sports' },
    { name: 'Slice of Life', stripe: 'bg-genre-slice-of-life' },
    { name: 'Horror', stripe: 'bg-genre-horror' },
    { name: 'Romance', stripe: 'bg-genre-romance' },
    { name: 'Comedy', stripe: 'bg-genre-comedy' },
    { name: 'Sci-Fi', stripe: 'bg-genre-science-fiction' },
    { name: 'Fantasy', stripe: 'bg-genre-fantasy' },
    { name: 'Action', stripe: 'bg-genre-action' },
    { name: 'Drama', stripe: 'bg-genre-drama' },
  ],

  search: {
    label: 'Search anime',
    placeholder: 'Search to add an anime — Frieren, Cowboy Bebop, Mushishi…',
    run: async (term) =>
      (await searchAnime(term)).map((anime) => {
        const [leading, other] = titleLines(anime);

        return {
          id: anime.id,
          title: leading,
          coverUrl: anime.coverUrl,
          // **Both lines are filled from the search itself**, unlike every other hobby here:
          // MAL answers the same node to a search and to a detail call, so nothing is waiting
          // on the title being added. The romaji name is under the English one, in the order
          // the card will show them, and the cour says which of several entries this is.
          byline: [other ?? '', seasonAndYear(anime.startSeason, anime.startYear) ?? ''],
          // No release window: MAL is not asked for one, so this hobby has no calendar for a
          // tile to offer. See HobbyDefinition.releases.
          release: null,
        };
      }),
  },

  journal: {
    load: async (mediaId) => {
      const anime = await getAnime(mediaId);
      const [leading, other] = titleLines(anime);

      return {
        // The card's order, and the drawer opens off the card: a heading that disagreed with
        // the thing just clicked would read as the wrong title having been opened.
        title: leading,
        subtitle: other,
        // The studio, which is what anybody names when they say who made an anime — a game's
        // developers and a film's directors answer the same question.
        byline: anime.studios,
        genres: anime.genres,
        primaryGenre: anime.primaryGenre,
        logEntries: anime.logEntries,
        platforms: [],
        // Empty, and that emptiness is the decision rather than an omission: a cour is its own
        // MAL entry, so there is no season to choose between.
        seasons: [],
        // What the one episode dropdown is sized from instead. Null for a cour that has not
        // aired — MAL answers nought there and nought means unknown, so the control offers
        // nothing rather than offering episode 0.
        episodeCount: anime.episodeCount,
        facts: animeFacts(anime),
        // MAL answers by id and the id came from MAL's own search, so there is no matcher to
        // have got it wrong — the film's argument, and what takes the pin off this drawer.
        hltb: null,
      };
    },

    setGenre: setAnimeGenre,
    setHltbId: null,

    // The episode alone, which is what makes this hobby's form different from every other one.
    // `ck_log_entries_episode_needs_season` forbade exactly this shape and was dropped for it.
    fields: { hoursPlayed: false, platform: false, progress: 'episode' },
  },

  progress: {
    // `E12`, and no season half to be missing. Television's `format` returns null for an
    // episode with no season, commenting that it would be inventing the half that is missing —
    // that comment is correct for television and this is why it is not a shared function.
    format: (_season, episode) => (episode === null ? null : `E${episode}`),

    describe: (_season, episode) =>
      episode === null ? 'Not started' : `Episode ${episode}`,
  },

  // No release calendar: this hobby's provider is not asked for a release window, so
  // nothing ever fills the columns the board would partition Backlog on. Turning it on is this
  // block plus that provider work, in one commit — see HobbyDefinition.releases.
  releases: null,
};

/**
 * A cour's two names, in the order they are read: the English one, then the romaji one.
 *
 * **The English name leads**, because it is what a person here calls the thing. `media.title`
 * still holds the romaji — that is MAL's own `title` and what its search matches on — so this
 * is a reading order rather than a second place a title is stored. The board row arrives with
 * the choice already made by the server, in `LibraryItemDto`; this is the same rule for the two
 * places that read a catalogue response directly, the search tile and the drawer.
 *
 * **The second name is dropped when it is the first one again.** Found by the e2e suite rather
 * than reasoned about, and not a stub artefact: MAL genuinely answers `alternative_titles.en`
 * of "Cowboy Bebop" for *Cowboy Bebop*, and does the same for every title whose romaji reading
 * is already English. Rendered blindly, those print their own name twice.
 *
 * Compared case-insensitively and with the ends trimmed, because "the same name" is a thing a
 * reader judges rather than a byte comparison — and **never any looser than that**, since
 * `Frieren` and `Frieren: Beyond Journey's End` are genuinely two names. `Card.tsx` states the
 * same rule over the row's own pair, which is the platform's copy of it rather than a duplicate
 * of this one: the card has to hold for a hobby that has not been written yet.
 */
function titleLines(anime: {
  title: string;
  englishTitle: string | null;
}): [leading: string, other: string | null] {
  const leading = anime.englishTitle ?? anime.title;
  const same = leading.trim().toLowerCase() === anime.title.trim().toLowerCase();

  return [leading, same ? null : anime.title];
}

/**
 * `Fall 2023`, which is how anybody says when a cour aired.
 *
 * Both halves or neither: a season with no year says nothing about *which* autumn, and a year
 * on its own is already the less interesting half. MAL's own word, capitalised — it sends
 * `fall` and nobody writes it that way.
 */
function seasonAndYear(season: string | null, year: number | null): string | null {
  if (year === null) {
    return null;
  }

  return season === null ? String(year) : `${capitalise(season)} ${year}`;
}

const capitalise = (word: string) => word.charAt(0).toUpperCase() + word.slice(1);

/**
 * The four facts about a cour, each dropped entirely when MAL has nothing to say.
 *
 * The films runtime rule: a dash there would read as a fact about the title rather than about
 * the request. Unlike a show, none of these arrives late — MAL answers everything to a search —
 * so a missing one is genuinely missing rather than not yet fetched.
 */
function animeFacts(anime: {
  externalId: string | null;
  mediaType: string | null;
  episodeCount: number | null;
  episodeRuntimeSeconds: number | null;
  airStatus: string | null;
  startSeason: string | null;
  startYear: number | null;
  sourceMaterial: string | null;
  meanScore: number | null;
}): TitleFact[] {
  const facts: TitleFact[] = [];

  const run = [
    anime.mediaType === null ? null : mediaTypeName(anime.mediaType),
    anime.episodeCount === null
      ? null
      : `${anime.episodeCount} episode${anime.episodeCount === 1 ? '' : 's'}`,
  ].filter((part) => part !== null);

  if (run.length > 0) {
    facts.push({ label: 'Run', value: run.join(' · ') });
  }

  // `Airing` and not `Status`, one word away from where *you* are on the title — the same
  // collision `air_status` avoids in the schema, and this is the reader-facing half of it.
  const airing = [
    anime.airStatus === null ? null : airStatusName(anime.airStatus),
    seasonAndYear(anime.startSeason, anime.startYear),
  ].filter((part) => part !== null);

  if (airing.length > 0) {
    facts.push({ label: 'Airing', value: airing.join(' · ') });
  }

  if (anime.episodeRuntimeSeconds !== null) {
    // One episode, in the same words a film's runtime is written in. The seconds are MAL's own
    // unit and are divided here rather than stored differently — the column keeps what the
    // provider said, and a display is the only place that has to pick a unit for a reader.
    facts.push({
      label: 'Episode',
      value: formatMinutes(Math.round(anime.episodeRuntimeSeconds / 60)),
    });
  }

  // MAL's own score, and the reason it is worth a line: it is out of ten, which is the scale a
  // pass's rating uses, so the two sit beside each other with no footnote.
  //
  // It was labelled `MAL` while that was the only thing here MAL's name could mean. It is not
  // any more — the link below wants the word more than the number does, because a label reading
  // `MAL` beside `View on MyAnimeList` is the one that explains itself. What says whose
  // opinion the score is, is that it sits in the band of things the provider states rather than
  // in the form below, where you record your own.
  if (anime.meanScore !== null) {
    facts.push({ label: 'Rating', value: anime.meanScore.toFixed(2) });
  }

  if (anime.sourceMaterial !== null) {
    // Whether a thing is an original or an adaptation is most of what anybody wants to know
    // before starting it.
    facts.push({ label: 'Source', value: sourceMaterialName(anime.sourceMaterial) });
  }

  // The way back to the entry this card was built from — last, because it is the only row here
  // that leaves the app, and everything above it is something to read rather than to press.
  //
  // **Not `source_lu.base_url`**, which is `https://api.myanimelist.net/v2/` — the API rather
  // than the site, and the two are not the same host. Written out here for the same reason
  // `HltbPin` writes howlongtobeat.com out: it is a fact about where a person goes, which is
  // the client's business, and threading it through the DTO would put a reader-facing URL in a
  // column that holds a machine-facing one.
  //
  // Dropped where there is no id to build it from. `external_id` is nullable because `media`'s
  // is, and an anime is only ever written by a MAL search — so this is a shape the app cannot
  // currently reach rather than one it reaches often. A link to `/anime/null` is worse than no
  // link, and this is the one fact here that is not MAL's opinion about the title: a cour MAL
  // knows nothing else about is exactly the one worth being able to open.
  if (anime.externalId !== null) {
    facts.push({
      label: 'MAL',
      value: 'View on MyAnimeList',
      href: `https://myanimelist.net/anime/${anime.externalId}`,
    });
  }

  return facts;
}

/**
 * MAL's `media_type`, in the words people use.
 *
 * `TV` and `OVA` and `ONA` are initialisms and stay upper-case; the rest are ordinary words.
 * Anything MAL adds later falls through as itself rather than as a blank — the value is theirs
 * to change, which is why the column is free text.
 */
function mediaTypeName(mediaType: string): string {
  const known: Record<string, string> = {
    tv: 'TV',
    tv_special: 'TV special',
    movie: 'Movie',
    ova: 'OVA',
    ona: 'ONA',
    special: 'Special',
    music: 'Music',
    cm: 'Commercial',
    pv: 'Promo',
  };

  return known[mediaType] ?? mediaType;
}

/** MAL's `status`, which is snake_case on the wire and a sentence to a reader. */
function airStatusName(status: string): string {
  const known: Record<string, string> = {
    finished_airing: 'Finished',
    currently_airing: 'Airing',
    not_yet_aired: 'Not yet aired',
  };

  return known[status] ?? status;
}

/** MAL's `source`, likewise — seventeen values, all snake_case, all theirs to change. */
function sourceMaterialName(source: string): string {
  const known: Record<string, string> = {
    original: 'Original',
    manga: 'Manga',
    '4_koma_manga': '4-koma manga',
    web_manga: 'Web manga',
    digital_manga: 'Digital manga',
    novel: 'Novel',
    light_novel: 'Light novel',
    web_novel: 'Web novel',
    visual_novel: 'Visual novel',
    game: 'Game',
    card_game: 'Card game',
    book: 'Book',
    picture_book: 'Picture book',
    radio: 'Radio',
    music: 'Music',
    mixed_media: 'Mixed media',
    other: 'Other',
  };

  return known[source] ?? source;
}
